### PR TITLE
Reset password

### PR DESCRIPTION
**Pull request by musaeed**

### CODE REVIEW

File: src/main/java/com/jersey/Authorization/Config/OAuth2Configuration.java [hunk 1]
```diff
@@ -67,10 +67,14 @@ public void configure(HttpSecurity http) throws Exception {
                 .and()
                 .authorizeRequests()
                 .antMatchers("/api/public/**").permitAll()
-                .antMatchers("/api/secure/**").authenticated();
-
+                .antMatchers("/api/secure/**").hasAnyAuthority(Authorities.ROLE_ADMIN.name(), Authorities.ROLE_USER.name(),Authorities.ROLE_USER_FACEBOOK.name())
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/Config/OAuth2Configuration.java:70]: shoudnt you add here CHANGE_PASSWORD_PRIVILEGE

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public void setExpiryDate(Date expiryDate) {
+        this.expiryDate = expiryDate;
+    }
+
+    private Date calculateExpiryDate() {
+
+        final Calendar cal = Calendar.getInstance();
+
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:49]: identation issues here only

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        cal.add(Calendar.MINUTE, EXPIRATION_TIME);
+
+        return new Date(cal.getTime().getTime());
+
+    }
+
+    @Override
+
+    public String toString() {
+
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:61]: identation ?

File: src/main/java/com/jersey/Authorization/security/PasswordResetTokenStore.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package com.jersey.Authorization.security;
+
+
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/security/PasswordResetTokenStore.java:3]: identation

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package com.jersey.Authorization.security;
+
+import java.util.Calendar;
+import java.util.Date;
+
+public class PasswordResetToken{
+
+    private static final int EXPIRATION_TIME = 60 * 24;
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:8]: so 24 minutes only. Why did you choose this number?

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    private String principal;
```
Comment (yar-malik) [inline src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:10]: what is principal here. I cant understand it intuitively. how would you describe it?

File: src/main/java/com/jersey/resources/EmailResource.java [hunk 7]
```diff
@@ -131,7 +130,7 @@ else if (email.locale.equalsIgnoreCase("de")){
         }
     }
 
-    public ClientResponse sendComplexMessage(String subject,
+    public static ClientResponse sendComplexMessage(String subject,
```
Comment (yar-malik) [inline src/main/java/com/jersey/resources/EmailResource.java:133]: why static? We dont want this. We want concurrent and parallelizaiton in programming

File: src/main/java/com/jersey/resources/EmailResource.java [hunk 8]
```diff
@@ -174,11 +173,40 @@ public ClientResponse sendComplexMessage(String subject,
...
+        formData.field("html", body);
+        formData.field("recipient-variables", recipientVariableJson.toString());
+
+        ClientResponse clientResponse = webResource.type(MediaType.MULTIPART_FORM_DATA_TYPE).post(ClientResponse.class, formData);
+        log.info(clientResponse.toString());
+
+        return clientResponse;
+    }
+
+    public static String htmlIntoString(String file) throws FileNotFoundException {
```
Comment (yar-malik) [inline src/main/java/com/jersey/resources/EmailResource.java:205]: static? Why?

File: src/main/java/com/jersey/resources/PersonResource.java [hunk 9]
```diff
@@ -245,6 +275,150 @@ public void delete(@PathParam("id") long id) {
...
+            emailTemplate = "en_reset_password.html";
+        }
+        else if(locale.equalsIgnoreCase("de")){
+            emailTemplate = "de_reset_password.html";
+        }
+
+        String redirectURL = new StringBuilder(backendURL)
+                            .append("/api/public/persons/changepassword")
+                            .append("?")
+                            .append("password_reset_token=").append(resetToken).append("&")
```
Comment (yar-malik) [inline src/main/java/com/jersey/resources/PersonResource.java:333]: is it safe to have this in URL. Will it be masked from other people.

File: src/main/java/com/jersey/resources/PersonResource.java [hunk 10]
```diff
@@ -245,6 +275,150 @@ public void delete(@PathParam("id") long id) {
...
+                            .append("user_email=").append(person.getEmail())
+                            .toString();
+
+        String html = EmailResource.htmlIntoString(emailTemplate);
+        html = html.replace("%password.resetlink%",redirectURL);
+
+        //send the email to the user
+        ClientResponse response = EmailResource.sendComplexMessage("Reset your password",
+                                                                   person.getEmail(),
+                                                        person.getFirstName()+ " " + person.getLastName(),
```
Comment (yar-malik) [inline src/main/java/com/jersey/resources/PersonResource.java:343]: indentation

File: src/main/java/com/jersey/resources/PersonResource.java [hunk 11]
```diff
@@ -245,6 +275,150 @@ public void delete(@PathParam("id") long id) {
...
+
+        String redirectURL = new StringBuilder(backendURL)
+                            .append("/api/public/persons/changepassword")
+                            .append("?")
+                            .append("password_reset_token=").append(resetToken).append("&")
+                            .append("user_email=").append(person.getEmail())
+                            .toString();
+
+        String html = EmailResource.htmlIntoString(emailTemplate);
+        html = html.replace("%password.resetlink%",redirectURL);
```
Comment (yar-malik) [inline src/main/java/com/jersey/resources/PersonResource.java:338]: does this work. Damn, I did things other way and this is definitely better.

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -6,5 +6,10 @@ authentication.oauth.clientid=webclient
 authentication.oauth.secret=secret
 authentication.oauth.tokenValidityInSeconds=43200
 
+#facebook credentials
 spring.social.facebook.appId=388810664881573
```
Comment (yar-malik) [inline src/main/resources/application.properties:10]: We cant commit this to code unfortunately. 
Anybody can use these secret keys unfortunately.

Comment (yar-malik) [reply src/main/resources/application.properties:10]: Check this answer. Exactly what you need here: 

https://stackoverflow.com/a/35535138/2359861

**Review by yar-malik — CHANGES_REQUESTED** [review]

File: src/main/java/com/jersey/Authorization/Config/OAuth2Configuration.java (hunk 1 above)
Comment (musaeed) [reply src/main/java/com/jersey/Authorization/Config/OAuth2Configuration.java:70]: no, we don't want users with change_password_privilege to access /api/secure api, only the /api/savepassword

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java (hunk 5 above)
Comment (musaeed) [reply src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:8]: 60 mins * 24 = 1 day

File: src/main/java/com/jersey/Authorization/security/PasswordResetToken.java (hunk 6 above)
Comment (musaeed) [reply src/main/java/com/jersey/Authorization/security/PasswordResetToken.java:10]: its the common nomenclature used by spring security, the user authenticated is always referred to as Principal, Look at the authentication interface:
https://docs.spring.io/spring-security/site/docs/3.0.x/apidocs/org/springframework/security/core/Authentication.html

File: src/main/java/com/jersey/resources/EmailResource.java (hunk 7 above)
Comment (musaeed) [reply src/main/java/com/jersey/resources/EmailResource.java:133]: static keyword doesn't have to do anything with parallelization or concurrency. It means that the method belongs to the class rather the instance, we don't want to create a new instance of the email resource everytime we want to send an email, that's just memory intensive. With static you can call method directly without creating instance, some scenerios when this should be used:
https://stackoverflow.com/questions/2671496/java-when-to-use-static-methods

File: src/main/java/com/jersey/resources/PersonResource.java (hunk 9 above)
Comment (musaeed) [reply src/main/java/com/jersey/resources/PersonResource.java:333]: its masked by mailgun, see the link when the email is recieved

File: src/main/resources/application.properties (hunk 12 above)
Comment (musaeed) [reply src/main/resources/application.properties:10]: i will fix this one plus the indentation ones

**Pull request closed without merging**